CustomElementQueue and CustomElementReactionQueue should have an inline capacity

rniwa opened the pull request:
#### 9b42484b8e812b6f398f55877d98a5a5845c7270
<pre>
CustomElementQueue and CustomElementReactionQueue should have an inline capacity
<a href="https://bugs.webkit.org/show_bug.cgi?id=258912">https://bugs.webkit.org/show_bug.cgi?id=258912</a>

Reviewed by Chris Dumez.

Both CustomElementQueue and CustomElementReactionQueue inline capacities of 4 and 1 respectively.

Also moved CustomElementReactionQueueItem from .cpp file to .h file and replaced some payload tuples
with structs to allow forward declarations of Document and HTMLFormElement.

* Source/WebCore/dom/CustomElementReactionQueue.cpp:
(WebCore::CustomElementReactionQueueItem::CustomElementReactionQueueItem):
(WebCore::CustomElementReactionQueueItem::invoke):
(WebCore::CustomElementReactionQueue::enqueueAdoptedCallbackIfNeeded):
(WebCore::CustomElementReactionQueue::enqueueFormAssociatedCallbackIfNeeded):
(WebCore::CustomElementReactionQueue::invokeAll):
(WebCore::CustomElementQueue::takeElements):
(WebCore::CustomElementReactionStack::takeElements):

* Source/WebCore/dom/CustomElementReactionQueue.h:
(WebCore::CustomElementReactionQueueItem): Moved from .cpp file.
(WebCore::CustomElementReactionQueueItem::type const):

* Source/WebCore/dom/GCReachableRef.h:
(WebCore::GCReachableRef::GCReachableRef): Removed the explicit requirement that T must be a subclass
of Node as this would require including Element.h in CustomElementReactionQueue.h. GCReachableRefMap
only supports Node anyway so this check isn&apos;t needed.

Canonical link: <a href="https://commits.webkit.org/265829@main">https://commits.webkit.org/265829@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/313e99dd3c58b80ffe37cf6fa09798d2f030e467

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12000 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12346 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12649 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13746 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11583 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12019 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14758 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12360 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14279 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12164 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12972 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10151 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14158 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10264 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10887 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18014 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11351 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11046 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14216 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11534 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9494 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10749 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15073 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1338 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11386 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/CustomElementReactionQueue.h:
```diff
@@ -40,29 +42,74 @@ class CallFrame;
+    void invoke(Element&, JSCustomElementInterface&);
+
+private:
+    Type m_type;
```
Seems we could use `{ Type::Invalid }` here.

rniwa replied:
Will do.

cdumez commented on Source/WebCore/dom/CustomElementReactionQueue.cpp:
```diff
@@ -42,84 +42,73 @@
+inline CustomElementReactionQueueItem::FormAssociatedPayload::~FormAssociatedPayload() = default;
+
+inline CustomElementReactionQueueItem::CustomElementReactionQueueItem()
+    : m_type(Type::Invalid)
```
If we added the inline initializer for the data member, we would `= default;` here.

rniwa replied:
Good point. Will do.

cdumez commented on Source/WebCore/dom/CustomElementReactionQueue.cpp:
```diff
@@ -42,84 +42,73 @@
+    : m_type(Type::Invalid)
+{ }
+
+inline CustomElementReactionQueueItem::CustomElementReactionQueueItem(CustomElementReactionQueueItem&& other)
```
Could we `= default;` here? Your implementation doesn't look different from what the compiler would generate for us.

rniwa replied:
Sure, will do

cdumez commented on Source/WebCore/dom/CustomElementReactionQueue.h:
```diff
@@ -40,29 +42,74 @@ class CallFrame;
+    WTF_MAKE_NONCOPYABLE(CustomElementReactionQueueItem);
+public:
+    enum class Type : uint8_t {
+        Invalid = 0,
```
`= 0` seems unnecessary

rniwa replied:
Will remove.

cdumez reviewed (approved):
LGTM

rniwa committed:
CustomElementQueue and CustomElementReactionQueue should have an inline capacity
https://bugs.webkit.org/show_bug.cgi?id=258912

Reviewed by Chris Dumez.

Both CustomElementQueue and CustomElementReactionQueue inline capacities of 4 and 1 respectively.

Also moved CustomElementReactionQueueItem from .cpp file to .h file and replaced some payload tuples
with structs to allow forward declarations of Document and HTMLFormElement.

* Source/WebCore/dom/CustomElementReactionQueue.cpp:
(WebCore::CustomElementReactionQueueItem::CustomElementReactionQueueItem):
(WebCore::CustomElementReactionQueueItem::invoke):
(WebCore::CustomElementReactionQueue::enqueueAdoptedCallbackIfNeeded):
(WebCore::CustomElementReactionQueue::enqueueFormAssociatedCallbackIfNeeded):
(WebCore::CustomElementReactionQueue::invokeAll):
(WebCore::CustomElementQueue::takeElements):
(WebCore::CustomElementReactionStack::takeElements):

* Source/WebCore/dom/CustomElementReactionQueue.h:
(WebCore::CustomElementReactionQueueItem): Moved from .cpp file.
(WebCore::CustomElementReactionQueueItem::type const):

* Source/WebCore/dom/GCReachableRef.h:
(WebCore::GCReachableRef::GCReachableRef): Removed the explicit requirement that T must be a subclass
of Node as this would require including Element.h in CustomElementReactionQueue.h. GCReachableRefMap
only supports Node anyway so this check isn't needed.

Canonical link: https://commits.webkit.org/265829@main
